### ISSUE TITLE
Layout: Adding legend for categorised vector with style copied from larger number categories

### ISSUE THREAD

**Issue by AdvisianADB** [qgis/QGIS#44665]
### What is the bug or the crash?

When adding and/or editting a legend item with categories based on selected values from a double type in a different vector file (copied all style items), QGIS crashes (V3.20.1, 3.16)

### Steps to reproduce the issue

Added a number of contour (line) shape files (consisting of integer isoline and double F (real 18 length, 16 precision)).  Values of F range from -6 to 50 in intervals of 2.  Selected values were from 2 to 50.  This style (all) was then copied to other contour files with smaller ranges of values.  When adding to a layout from one of the contour shape files (with a smaller range) QGIS hung and then crashed.

### Versions

QGIS version
3.20.1-Odense
QGIS code revision
1c3c5cd6
Qt version
5.15.2
Python version
3.9.5
GDAL/OGR version
3.3.1
PROJ version
8.1.0
EPSG Registry database version
v10.018 (2021-04-02)
GEOS version
3.9.1-CAPI-1.14.2
SQLite version
3.35.2
PDAL version
2.3.0
PostgreSQL client version
13.0
SpatiaLite version
5.0.1
QWT version
6.1.3
QScintilla2 version
2.11.5
OS version
Windows 10 Version 1909

Active Python plugins
DataPlotly
FreehandRasterGeoreferencer
geoscience
geospatialsimulation
midvatten
mmqgis
pointsamplingtool
processing_fusion
qgis-maptiler-plugin
Qgis2threejs
quick_map_services
rastertimeseriesmanager
Serval
slyr_community
tuflow
db_manager
MetaSearch
processing

## User Feedback

Adding legend for item with style which includes items/categories that are not present in current shape file (style copied from another shape file which includes all categories)

## Report Details

**Crash ID**: [8b7a3c016f7fda2d1464578c601c8e677165a67c](https://github.com/qgis/QGIS/search?q=8b7a3c016f7fda2d1464578c601c8e677165a67c&type=Issues)

**Stack Trace**
<pre>
QgsLayerTreeModel::node2index :
QgsLayerTreeView::updateExpandedStateFromNode :
QgsLayerTreeView::modelRowsInserted :
QObject::qt_static_metacall :
QAbstractItemModel::endInsertRows :
QSortFilterProxyModel::insertRows :
QSortFilterProxyModel::sortRoleChanged :
QSortFilterProxyModel::qt_static_metacall :
QObject::qt_static_metacall :
QAbstractItemModel::endInsertRows :
QgsLayerTreeModel::addLegendToLayer :
QgsLayerTreeModel::connectToLayer :
QgsLayerTreeModel::nodeAddedChildren :
QObject::qt_static_metacall :
QgsLayerTreeNode::addedChildren :
QgsLayerTreeNode::insertChildrenPrivate :
QgsLayerTreeGroup::insertChildNodes :
QgsLayerTreeGroup::insertChildNode :
QgsLayerTreeGroup::addLayer :
QgsLayoutLegendWidget::mAddToolButton_clicked :
QObject::qt_static_metacall :
QAbstractButton::clicked :
QAbstractButton::click :
QAbstractButton::mouseReleaseEvent :
QToolButton::mouseReleaseEvent :
QWidget::event :
QApplicationPrivate::notify_helper :
QApplication::notify :
QgsApplication::notify :
QCoreApplication::notifyInternal2 :
QApplicationPrivate::sendMouseEvent :
QSizePolicy::QSizePolicy :
QSizePolicy::QSizePolicy :
QApplicationPrivate::notify_helper :
QApplication::notify :
QgsApplication::notify :
QCoreApplication::notifyInternal2 :
QGuiApplicationPrivate::processMouseEvent :
QWindowSystemInterface::sendWindowSystemEvents :
QEventDispatcherWin32::processEvents :
qt_plugin_query_metadata :
QEventLoop::exec :
QCoreApplication::exec :
main :
BaseThreadInitThunk :
RtlUserThreadStart :
</pre>

**QGIS Info**
QGIS Version: 3.20.1-Odense
QGIS code revision: 1c3c5cd6
Compiled against Qt: 5.15.2
Running against Qt: 5.15.2
Compiled against GDAL: 3.3.1
Running against GDAL: 3.3.1

**System Info**
CPU Type: x86_64
Kernel Type: winnt
Kernel Version: 10.0.18363

**Comment by gioman** [comment]
@AdvisianADB can you please attach sample data and styles? Have you tried on a new QGIS profile?

**Comment by AdvisianADB** [comment]
Hi Giovanni,
I’ve attached a zip file containing the data files and a cut-down QGIS file (LabelProblem).  I did not have any problems with constructing the layout.
I also tried to reconstruct the problem (LabelProblemV2 and LabelProblemV2B where in 2B I recreated the steps (partially) in relabeling the categories (lines) in Y01_4B_DrawdownS7 after copying the style (all style categories) from Y50_4B_DrawdownS7).  However instead of crashing this time, QGIS updated the legend on the print layout.
Regards
Tony
